akka-stream/src/main/scala/akka/stream/StreamRefs.scala

```diff
@@ -0,0 +1,78 @@
+package akka.stream
+
+import akka.actor.{ ActorRef, DeadLetterSuppression }
+import akka.annotation.InternalApi
+import akka.stream.impl.ReactiveStreamsCompliance
+
+/**
+ * INTERNAL API: Use [[akka.stream.scaladsl.SourceRef]] and [[akka.stream.scaladsl.SinkRef]] directly to obtain stream refs.
+ */
+@InternalApi
+private[akka] object StreamRefs {
+
+  @InternalApi
+  sealed trait Protocol
+
+  /**
+   * Sequenced `Subscriber#onNext` equivalent.
+   * The receiving end of these messages MUST fail the stream if it observes gaps in the sequence,
+   * as these messages will not be re-delivered.
+   *
+   * Sequence numbers start from `0`.
+   */
+  @InternalApi
+  final case class SequencedOnNext[T](seqNr: Long, payload: T) extends StreamRefs.Protocol {
+    if (payload == null) throw ReactiveStreamsCompliance.elementMustNotBeNullException
+  }
+
+  final case class OnSubscribeHandshake(targetRef: ActorRef) extends StreamRefs.Protocol
+
+  /** Sent to a the receiver side of a stream ref, once the sending side of the SinkRef gets signalled a Failure. */
+  @InternalApi
+  final case class RemoteStreamFailure(msg: String) extends StreamRefs.Protocol
+
+  /** Sent to a the receiver side of a stream ref, once the sending side of the SinkRef gets signalled a completion. */
+  @InternalApi
+  final case class RemoteStreamCompleted(seqNr: Long) extends StreamRefs.Protocol
+
+  /**
+   * Cumulative demand, equivalent to sequence numbering all events in a stream. *
+   * This message may be re-delivered.
+   */
+  @InternalApi
+  final case class CumulativeDemand(seqNr: Long) extends StreamRefs.Protocol with DeadLetterSuppression {
+    if (seqNr <= 0) throw ReactiveStreamsCompliance.numberOfElementsInRequestMustBePositiveException
+  }
+
+  // --- exceptions ---
+
+  final case class TargetRefNotInitializedYetException()
```

Comment:
If these are thought to be public API then they must be moved out of the `private` object.

Reply:
Good catch!